File: snippets/algorithms/graph-traversal/graph_traversal.rb
#!/usr/bin/env ruby

# Graph Traversal Algorithms in Ruby
#
# This script demonstrates BFS and DFS traversal algorithms on a graph.
#
# Run: ruby graph_traversal.rb

class Graph
  # Initialize an empty graph
  def initialize
    @adjacency_list = {}
  end

  # Add a vertex to the graph
  def add_vertex(vertex)
    @adjacency_list[vertex] = [] unless @adjacency_list.key?(vertex)
  end

  # Add an edge between two vertices
  def add_edge(v1, v2)
    # Ensure both vertices exist
    add_vertex(v1)
    add_vertex(v2)
    
    # Add the edge (undirected graph)
    @adjacency_list[v1] << v2
    @adjacency_list[v2] << v1
  end

  # Get sorted neighbors for consistent output
  def get_sorted_neighbors(vertex)
    @adjacency_list[vertex].sort
  end

  # Breadth-First Search traversal
  def bfs(start)
    return [] unless @adjacency_list.key?(start)
    
    visited = { start => true }
    queue = [start]
    result = []
    
    puts "Starting BFS traversal from vertex #{start}"
    
    until queue.empty?
      # Dequeue the first vertex
      vertex = queue.shift
      result << vertex
      
      puts "Visiting: #{vertex}"
      puts "Queue: #{queue.inspect}"
      puts "Visited so far: #{result.inspect}"
      puts "------------------------------"
      
      # Pause for demonstration
      sleep(0.5) # 500ms
      
      # Get sorted neighbors for consistent order
      neighbors = get_sorted_neighbors(vertex)
      
      # Enqueue all unvisited neighbors
      neighbors.each do |neighbor|
        unless visited[neighbor]
          visited[neighbor] = true
          queue << neighbor
        end
      end
    end
    
    result
  end

  # Depth-First Search traversal (recursive)
  def dfs_recursive(start)
    return [] unless @adjacency_list.key?(start)
    
    visited = {}
    result = []
    
    puts "Starting recursive DFS traversal from vertex #{start}"
    
    dfs_helper(start, visited, result)
    
    result
  end

  # Helper method for recursive DFS
  def dfs_helper(vertex, visited, result)
    # Mark as visited and add to result
    visited[vertex] = true
    result << vertex
    
    puts "Visiting: #{vertex}"
    puts "Visited so far: #{result.inspect}"
    puts "------------------------------"
    
    # Pause for demonstration
    sleep(0.5) # 500ms
    
    # Get sorted neighbors for consistent order
    neighbors = get_sorted_neighbors(vertex)
    
    # Recursively visit all unvisited neighbors
    neighbors.each do |neighbor|
      dfs_helper(neighbor, visited, result) unless visited[neighbor]
    end
  end

  # Depth-First Search traversal (iterative)
  def dfs_iterative(start)
    return [] unless @adjacency_list.key?(start)
    
    visited = {}
    stack = [start]
    result = []
    
    puts "Starting iterative DFS traversal from vertex #{start}"
    
    until stack.empty?
      # Pop the top vertex
      vertex = stack.pop
      
      # If not visited, process it
      unless visited[vertex]
        visited[vertex] = true
        result << vertex
        
        puts "Visiting: #{vertex}"
        puts "Stack: #{stack.inspect}"
        puts "Visited so far: #{result.inspect}"
        puts "------------------------------"
        
        # Pause for demonstration
        sleep(0.5) # 500ms
        
        # Get sorted neighbors in reverse order for stack
        neighbors = get_sorted_neighbors(vertex).reverse
        
        # Push all unvisited neighbors onto the stack
        neighbors.each do |neighbor|
          stack << neighbor unless visited[neighbor]
        end
      end
    end
    
    result
  end

  # Print the graph structure
  def visualize_graph
    puts "\nGraph Structure:"
    puts "------------------------------"
    
    # Sort vertices for consistent output
    vertices = @adjacency_list.keys.sort
    
    vertices.each do |vertex|
      neighbors = get_sorted_neighbors(vertex)
      puts "#{vertex} -> #{neighbors.inspect}"
    end
    
    puts "------------------------------"
  end
end

# Create a sample graph for demonstration
def create_sample_graph
  g = Graph.new
  
  # Add edges to build this graph:
  #     A
  #    / \
  #   B   C
  #  / \   \
  # D   E---F
  
  edges = [
    ["A", "B"], ["A", "C"],
    ["B", "D"], ["B", "E"],
    ["C", "F"], ["E", "F"]
  ]
  
  edges.each do |v1, v2|
    g.add_edge(v1, v2)
  end
  
  g
end

# Main execution
g = create_sample_graph
g.visualize_graph

# Demonstrate BFS
puts "\n=== BFS Traversal ==="
bfs_result = g.bfs("A")
puts "BFS Result: #{bfs_result.inspect}"

# Demonstrate recursive DFS
puts "\n=== DFS Traversal (Recursive) ==="
dfs_rec_result = g.dfs_recursive("A")
puts "DFS Recursive Result: #{dfs_rec_result.inspect}"

# Demonstrate iterative DFS
puts "\n=== DFS Traversal (Iterative) ==="
dfs_iter_result = g.dfs_iterative("A")
puts "DFS Iterative Result: #{dfs_iter_result.inspect}" 

File: snippets/algorithms/graph-traversal/graph_traversal.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::thread;
use std::time::Duration;

/// Graph Traversal Algorithms in Rust
///
/// This program demonstrates BFS and DFS traversal algorithms on a graph.
///
/// Compile: rustc graph_traversal.rs
/// Run: ./graph_traversal

/// A graph using adjacency list representation
struct Graph {
    // Adjacency list representation
    adjacency_list: HashMap<String, Vec<String>>,
}

impl Graph {
    /// Creates a new empty graph
    fn new() -> Self {
        Graph {
            adjacency_list: HashMap::new(),
        }
    }

    /// Adds a vertex to the graph
    fn add_vertex(&mut self, vertex: &str) {
        self.adjacency_list.entry(vertex.to_string()).or_insert(Vec::new());
    }

    /// Adds an edge between two vertices
    fn add_edge(&mut self, v1: &str, v2: &str) {
        // Ensure both vertices exist
        self.add_vertex(v1);
        self.add_vertex(v2);
        
        // Add the edge (undirected graph)
        self.adjacency_list.get_mut(v1).unwrap().push(v2.to_string());
        self.adjacency_list.get_mut(v2).unwrap().push(v1.to_string());
    }

    /// Helper method to get sorted neighbors for consistent output
    fn get_sorted_neighbors(&self, vertex: &str) -> Vec<String> {
        let mut neighbors = self.adjacency_list[vertex].clone();
        neighbors.sort();
        neighbors
    }

    /// Performs a breadth-first search traversal starting from the given vertex
    fn bfs(&self, start: &str) -> Vec<String> {
        if !self.adjacency_list.contains_key(start) {
            return Vec::new();
        }
        
        let mut visited = HashSet::new();
        let mut queue = VecDeque::new();
        let mut result = Vec::new();
        
        // Initialize with starting vertex
        visited.insert(start.to_string());
        queue.push_back(start.to_string());
        
        println!("Starting BFS traversal from vertex {}", start);
        
        while !queue.is_empty() {
            // Dequeue the first vertex
            let vertex = queue.pop_front().unwrap();
            result.push(vertex.clone());
            
            println!("Visiting: {}", vertex);
            println!("Queue: {:?}", queue);
            println!("Visited so far: {:?}", result);
            println!("------------------------------");
            
            // Pause for demonstration
            thread::sleep(Duration::from_millis(500));
            
            // Get sorted neighbors for consistent order
            let neighbors = self.get_sorted_neighbors(&vertex);
            
            // Enqueue all unvisited neighbors
            for neighbor in neighbors {
                if !visited.contains(&neighbor) {
                    visited.insert(neighbor.clone());
                    queue.push_back(neighbor);
                }
            }
        }
        
        result
    }

    /// Performs a recursive depth-first search traversal starting from the given vertex
    fn dfs_recursive(&self, start: &str) -> Vec<String> {
        if !self.adjacency_list.contains_key(start) {
            return Vec::new();
        }
        
        let mut visited = HashSet::new();
        let mut result = Vec::new();
        
        println!("Starting recursive DFS traversal from vertex {}", start);
        
        self.dfs_helper(start, &mut visited, &mut result);
        
        result
    }

    /// Helper method for recursive DFS
    fn dfs_helper(&self, vertex: &str, visited: &mut HashSet<String>, result: &mut Vec<String>) {
        // Mark as visited and add to result
        visited.insert(vertex.to_string());
        result.push(vertex.to_string());
        
        println!("Visiting: {}", vertex);
        println!("Visited so far: {:?}", result);
        println!("------------------------------");
        
        // Pause for demonstration
        thread::sleep(Duration::from_millis(500));
        
        // Get sorted neighbors for consistent order
        let neighbors = self.get_sorted_neighbors(vertex);
        
        // Recursively visit all unvisited neighbors
        for neighbor in neighbors {
            if !visited.contains(&neighbor) {
                self.dfs_helper(&neighbor, visited, result);
            }
        }
    }

    /// Performs an iterative depth-first search traversal starting from the given vertex
    fn dfs_iterative(&self, start: &str) -> Vec<String> {
        if !self.adjacency_list.contains_key(start) {
            return Vec::new();
        }
        
        let mut visited = HashSet::new();
        let mut stack = Vec::new();
        let mut result = Vec::new();
        
        // Initialize with starting vertex
        stack.push(start.to_string());
        
        println!("Starting iterative DFS traversal from vertex {}", start);
        
        while !stack.is_empty() {
            // Pop the top vertex
            let vertex = stack.pop().unwrap();
            
            // If not visited, process it
            if !visited.contains(&vertex) {
                visited.insert(vertex.clone());
                result.push(vertex.clone());
                
                println!("Visiting: {}", vertex);
                println!("Stack: {:?}", stack);
                println!("Visited so far: {:?}", result);
                println!("------------------------------");
                
                // Pause for demonstration
                thread::sleep(Duration::from_millis(500));
                
                // Get sorted neighbors in reverse order for stack
                let mut neighbors = self.get_sorted_neighbors(&vertex);
                neighbors.reverse();
                
                // Push all unvisited neighbors onto the stack
                for neighbor in neighbors {
                    if !visited.contains(&neighbor) {
                        stack.push(neighbor);
                    }
                }
            }
        }
        
        result
    }

    /// Prints a visualization of the graph structure
    fn visualize_graph(&self) {
        println!("\nGraph Structure:");
        println!("------------------------------");
        
        // Sort vertices for consistent output
        let mut vertices: Vec<String> = self.adjacency_list.keys().cloned().collect();
        vertices.sort();
        
        for vertex in vertices {
            let neighbors = self.get_sorted_neighbors(&vertex);
            println!("{} -> {:?}", vertex, neighbors);
        }
        
        println!("------------------------------");
    }
}

/// Creates a sample graph for demonstration
fn create_sample_graph() -> Graph {
    let mut g = Graph::new();
    
    // Add edges to build this graph:
    //     A
    //    / \
    //   B   C
    //  / \   \
    // D   E---F
    
    let edges = [
        ("A", "B"), ("A", "C"),
        ("B", "D"), ("B", "E"),
        ("C", "F"), ("E", "F")
    ];
    
    for (v1, v2) in edges.iter() {
        g.add_edge(v1, v2);
    }
    
    g
}

fn main() {
    // Create a sample graph
    let g = create_sample_graph();
    g.visualize_graph();
    
    // Demonstrate BFS
    println!("\n=== BFS Traversal ===");
    let bfs_result = g.bfs("A");
    println!("BFS Result: {:?}", bfs_result);
    
    // Demonstrate recursive DFS
    println!("\n=== DFS Traversal (Recursive) ===");
    let dfs_rec_result = g.dfs_recursive("A");
    println!("DFS Recursive Result: {:?}", dfs_rec_result);
    
    // Demonstrate iterative DFS
    println!("\n=== DFS Traversal (Iterative) ===");
    let dfs_iter_result = g.dfs_iterative("A");
    println!("DFS Iterative Result: {:?}", dfs_iter_result);
} 